src/graph/dfs.rs

```diff
@@ -0,0 +1,121 @@
+use super::Graph;
+use crate::graph::AdjListIterator;
+use bit_vec::BitVec;
+
+impl Graph {
+    pub fn dfs(&self, v: usize) -> DfsIterator {
+        // Create a stack for DFS
+        let mut stack: Vec<usize> = Vec::new();
+
+        let adj_iters = (0..self.num_v())
+            .map(|u| self.adj_list(u))
+            .collect::<Vec<_>>();
+
+        // Push the current source node.
+        stack.push(v);
+
+        DfsIterator {
+            visited: BitVec::from_elem(self.num_v(), false),
+            stack,
+            adj_iters,
+        }
+    }
+}
+pub struct DfsIterator<'a> {
+    //is vertex visited
+    visited: BitVec,
+    //stack of vertices
+    stack: Vec<usize>,
+    adj_iters: Vec<AdjListIterator<'a>>,
+}
+
+impl<'a> Iterator for DfsIterator<'a> {
+    type Item = usize;
+
+    /// Returns next vertex in the DFS
+    fn next(&mut self) -> Option<Self::Item> {
+        //Sources:
+        // https://www.geeksforgeeks.org/iterative-depth-first-traversal/
+        // https://en.wikipedia.org/wiki/Depth-first_search
+        while let Some(s) = self.stack.last() {
+            let s = *s;
+
+            //Does s still have neighbors we need to process?
+            if let Some(s_nbr) = self.adj_iters[s].next() {
```

Comment:
Some((_, v)) can be used to extract the vertex directly

Reply:
Yes, it is cleaner that way, done